File: src/types/remediation.ts
export interface RemediationExecutionLog {
  log_id: string;
  investigation_id: string;
  action_id: string;
  target_system: string;
  status: 'SUCCESS' | 'FAILED' | 'REVERTED' | 'RUNNING' | 'PENDING' | 'SIMULATED';
  is_dry_run: boolean;
  affected_indicators: string[];
  execution_result: {
    status?: string;
    action?: string;
    target_system?: string;
    confirmation_id?: string;
    message?: string;
    rollback_supported?: boolean;
    rollback_token?: string;
    rollback_status?: string;
    [key: string]: any;
  };
  executed_by: string;
  executed_at: string;
  reverted_at?: string | null;
  rollback_supported: boolean;
}

export interface RemediationHistoryResponse {
  investigation_id: string;
  total_executions: number;
  active_enforcements: number;
  logs: RemediationExecutionLog[];
}

export interface RemediationExecuteRequest {
  action_id: string;
  target_system?: string;
  action_payload?: Record<string, any>;
  dry_run?: boolean;
}

export interface RemediationBatchRequest {
  action_ids?: string[];
  priority_filter?: string;
  dry_run?: boolean;
}

export interface STIXObject {
  type: string;
  spec_version?: string;
  id: string;
  name?: string;
  description?: string;
  created?: string;
  modified?: string;
  [key: string]: any;
}

export interface STIXBundle {
  type: 'bundle';
  id: string;
  objects: STIXObject[];
}

File: backend/app/schemas/remediation.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class RemediationExecuteRequest(BaseModel):
    action_id: str = Field(..., description="Action ID from DFIR report (e.g. ACT-01, ACT-02)")
    target_system: Optional[str] = Field(None, description="Target enforcement system (DNS, SWG, MTA, Exchange, EDR)")
    action_payload: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Custom payload parameters or indicator overrides")
    dry_run: bool = Field(default=False, description="If True, simulates execution without making external API changes")


class RemediationBatchRequest(BaseModel):
    action_ids: Optional[List[str]] = Field(default=None, description="List of specific action IDs to execute. If omitted, executes all matching priority_filter.")
    priority_filter: Optional[str] = Field(default="P0", description="Priority tier to execute (e.g. P0, P1, P2)")
    dry_run: bool = Field(default=False, description="If True, simulates execution without making external API changes")


class RemediationExecutionResponse(BaseModel):
    log_id: str
    investigation_id: str
    action_id: str
    target_system: str
    status: str = Field(..., description="SUCCESS, FAILED, REVERTED, RUNNING, PENDING")
    is_dry_run: bool = False
    affected_indicators: List[str] = Field(default_factory=list)
    execution_result: Dict[str, Any] = Field(default_factory=dict)
    executed_by: str = "usr-analyst-001"
    executed_at: str
    reverted_at: Optional[str] = None
    rollback_supported: bool = True


class RemediationHistoryResponse(BaseModel):
    investigation_id: str
    total_executions: int
    active_enforcements: int
    logs: List[RemediationExecutionResponse] = Field(default_factory=list)


class STIXBundleDTO(BaseModel):
    type: str = "bundle"
    id: str
    objects: List[Dict[str, Any]] = Field(default_factory=list)
